lib/teiserver/battle/balance/brute_force_avoid.ex

```diff
@@ -0,0 +1,161 @@
+defmodule Teiserver.Battle.Balance.BruteForceAvoid do
+  @moduledoc """
+  Overview:
+  Go through every possible combination and pick the best one. Each combination is given a score
+  Score = team rating penalty + broken avoid penalty + broken party penalty + broken max team rating diff penalty
+  team rating penalty = difference in team ratings
+  broken avoid penalty = broken avoid count * avoid importance
+  broken party penalty = broken party count * party importance
+  broken max team rating diff penalty = max team diff importance (if the team rating diff > @max_team_diff)
+
+  Only use for games with two teams.
+
+  This is not a balance algorithm that is callable players. It's a lib that can be used by another balance algorithm.
+  """
+  alias Teiserver.Battle.Balance.BruteForceAvoidTypes, as: BF
+  require Integer
+
+  @max_team_diff 17
+  @max_team_diff_importance 10000
+  @party_importance 1000
+  @avoid_importance 10
+
+  def get_best_combo(players, avoids, parties) do
+    potential_teams = potential_teams(length(players))
+    get_best_combo(potential_teams, players, avoids, parties)
+  end
+
+  @spec get_best_combo([integer()], [BF.player()], [[number()]], [[number()]]) ::
+          BF.combo_result()
+  def get_best_combo(combos, players, avoids, parties) do
+    players_with_index = Enum.with_index(players)
+
+    # Go through every possibility and get the combination with the lowest score
+    result =
+      Enum.map(combos, fn x ->
+        get_players_from_indexes(x, players_with_index)
+      end)
+      |> Enum.map(fn team ->
+        result = score_combo(team, players, avoids, parties)
+        Map.put(result, :first_team, team)
+      end)
+      |> Enum.min_by(fn z ->
+        z.score
+      end)
+
+    first_team = result.first_team
+
+    second_team =
+      players
+      |> Enum.filter(fn x ->
+        !Enum.any?(first_team, fn y ->
+          y.id == x.id
+        end)
+      end)
+
+    Map.put(result, :second_team, second_team)
+  end
+
+  @spec potential_teams(integer()) :: any()
+  def potential_teams(num_players) do
+    Teiserver.Helper.CombinationsHelper.get_combinations(num_players)
+  end
+
+  @spec score_combo([BF.player()], [BF.player()], [[number()]], [[number()]]) :: any()
+  def score_combo(first_team, all_players, avoids, parties) do
+    first_team_rating = get_team_rating(first_team)
+    both_team_rating = get_team_rating(all_players)
+    rating_diff_penalty = abs(both_team_rating - first_team_rating * 2)
+
+    max_team_diff_penalty =
+      cond do
+        rating_diff_penalty > @max_team_diff -> @max_team_diff_importance
```

Comment:
Why use some hard coded value instead of open skill win prediction? Seems to me that hardcoded value is very binary and having a single value for all team sizes also sounds off.

Reply:
We don't have access to win prediction in Elixir lib. Hence why I asked on discord for this:
https://ptb.discord.com/channels/549281623154229250/855772061095559179/1276661091346153533

Reply:
Calling API to get win prediction as part of balancing when you are checking _many_ combinations and need score for each: very bad idea.

https://github.com/vivekjoshy/openskill.py/blob/main/openskill/models/weng_lin/plackett_luce.py#L846-L858 looks very easy to port.

I'm not asking about simply dividing the sum. But also what you do at the moment, it's arbitrary cutoff: < 17: all good max score, > 17: 0 score. For example with this as far as I understand the core, balancing [[24,24], [15,18]] will be just fine and get max score, that sounds very wrong to me.

Reply:
I'll take a look at if we can port it.

Reply:
The addition of win probability to Elixir library will be done in a seperate PR. This PR will just calculate team ratings as before.